network/netcore/src/multiplexing/yamux.rs

```diff
@@ -76,152 +67,127 @@ where
-                io::ErrorKind::Other,
-                "Unable to open substream; underlying connection is dead",
-            )),
-            Err(e) => Err(io::Error::new(io::ErrorKind::Other, e)),
-        };
-
-        future::ready(output)
-    }
-
-    fn close(&self) -> Self::Close {
-        Close::new(self.inner.clone())
-    }
-}
-
-#[derive(Debug)]
-#[must_use = "futures do nothing unless polled"]
-pub struct Close<TSocket> {
-    inner: yamux::Connection<Compat<TSocket>>,
+#[derive(Clone)]
+pub struct YamuxControl {
+    inner: yamux::Control,
 }
 
-impl<TSocket> Close<TSocket>
-where
-    TSocket: AsyncRead + AsyncWrite + Unpin,
-{
-    pub fn new(connection: yamux::Connection<Compat<TSocket>>) -> Self {
-        Self { inner: connection }
-    }
-}
+#[async_trait]
+impl Control for YamuxControl {
+    type Substream = StreamHandle;
 
-impl<TSocket> Future for Close<TSocket>
-where
-    TSocket: AsyncRead + AsyncWrite + Unpin,
-{
-    type Output = io::Result<()>;
-
-    fn poll(self: Pin<&mut Self>, context: &mut Context) -> Poll<Self::Output> {
-        let mut close_fut = Compat01As03::new(poll_fn_01(|| {
-            self.inner
-                .close()
-                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
-        }));
-        Pin::new(&mut close_fut).poll(context)
+    async fn open_stream(&mut self) -> io::Result<Self::Substream> {
+        self.inner
+            .open_stream()
+            .await
+            .map_err(|e| io::Error::new(io::ErrorKind::BrokenPipe, e))
     }
-}
-
-#[derive(Debug)]
-#[must_use = "streams do nothing unless polled"]
-pub struct Listener<TSocket> {
-    inner: Compat01As03<yamux::Connection<Compat<TSocket>>>,
-}
 
-impl<TSocket> Listener<TSocket>
-where
-    TSocket: AsyncRead + AsyncWrite + Unpin,
-{
-    pub fn new(connection: yamux::Connection<Compat<TSocket>>) -> Self {
-        Self {
-            inner: Compat01As03::new(connection),
-        }
+    async fn close(&mut self) -> io::Result<()> {
+        self.inner
+            .close()
+            .await
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
     }
 }
 
-impl<TSocket> Stream for Listener<TSocket>
+#[async_trait]
+impl<TSocket> StreamMultiplexer for Yamux<TSocket>
 where
-    TSocket: AsyncRead + AsyncWrite + Unpin,
+    TSocket: AsyncRead + AsyncWrite + Send + Debug + Unpin + Sync + 'static,
 {
-    type Item = io::Result<Compat01As03<yamux::StreamHandle<Compat<TSocket>>>>;
-
-    fn poll_next(mut self: Pin<&mut Self>, context: &mut Context) -> Poll<Option<Self::Item>> {
-        match Pin::new(&mut self.inner).poll_next(context) {
-            Poll::Pending => Poll::Pending,
-            Poll::Ready(None) => Poll::Ready(None),
-            Poll::Ready(Some(Err(e))) => {
-                Poll::Ready(Some(Err(io::Error::new(io::ErrorKind::Other, e))))
+    type Substream = StreamHandle;
+    type Control = YamuxControl;
+    type Listener = BoxStream<'static, io::Result<StreamHandle>>;
```

Comment:
does this work?

`type Listener = mpsc::Receiver<io::Result<StreamHandle>>` then return `(rx, YamuxControl { inner: control })` below?

I'm guessing there is some weird trait bound stuff?

Reply:
No, this works fine!